File: src/main/java/com/owncloud/android/ui/activity/UploadListActivity.java
/*
 * Nextcloud Android client application
 *
 * @author Tobias Kaminsky
 * @author Chris Narkiewicz
 * Copyright (C) 2018 Tobias Kaminsky
 * Copyright (C) 2018 Nextcloud
 * Copyright (C) 2019 Chris Narkiewicz <hello@ezaquarii.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU AFFERO GENERAL PUBLIC LICENSE
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU AFFERO GENERAL PUBLIC LICENSE for more details.
 *
 * You should have received a copy of the GNU Affero General Public
 * License along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.owncloud.android.ui.activity;

import android.accounts.Account;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.IBinder;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;

import com.evernote.android.job.Job;
import com.evernote.android.job.JobManager;
import com.evernote.android.job.JobRequest;
import com.evernote.android.job.util.support.PersistableBundleCompat;
import com.nextcloud.client.account.User;
import com.nextcloud.client.account.UserAccountManager;
import com.nextcloud.client.core.Clock;
import com.nextcloud.client.device.PowerManagementService;
import com.nextcloud.client.network.ConnectivityService;
import com.nextcloud.java.util.Optional;
import com.owncloud.android.R;
import com.owncloud.android.databinding.UploadListLayoutBinding;
import com.owncloud.android.datamodel.UploadsStorageManager;
import com.owncloud.android.files.services.FileUploader;
import com.owncloud.android.files.services.FileUploader.FileUploaderBinder;
import com.owncloud.android.jobs.FilesSyncJob;
import com.owncloud.android.lib.common.operations.RemoteOperation;
import com.owncloud.android.lib.common.operations.RemoteOperationResult;
import com.owncloud.android.lib.common.utils.Log_OC;
import com.owncloud.android.operations.CheckCurrentCredentialsOperation;
import com.owncloud.android.ui.adapter.UploadListAdapter;
import com.owncloud.android.ui.decoration.MediaGridItemDecoration;
import com.owncloud.android.utils.FilesSyncHelper;
import com.owncloud.android.utils.ThemeUtils;

import java.util.Set;

import javax.inject.Inject;

import androidx.recyclerview.widget.GridLayoutManager;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

/**
 * Activity listing pending, active, and completed uploads. User can delete
 * completed uploads from view. Content of this list of coming from
 * {@link UploadsStorageManager}.
 */
public class UploadListActivity extends FileActivity {

    private static final String TAG = UploadListActivity.class.getSimpleName();

    private UploadMessagesReceiver uploadMessagesReceiver;

    private UploadListAdapter uploadListAdapter;

    public SwipeRefreshLayout swipeListRefreshLayout;

    @Inject
    UserAccountManager userAccountManager;

    @Inject
    UploadsStorageManager uploadsStorageManager;

    @Inject
    ConnectivityService connectivityService;

    @Inject
    PowerManagementService powerManagementService;

    @Inject
    Clock clock;

    private UploadListLayoutBinding binding;

    @Override
    public void showFiles(boolean onDeviceOnly) {
        super.showFiles(onDeviceOnly);
        Intent i = new Intent(getApplicationContext(), FileDisplayActivity.class);
        i.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
        startActivity(i);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        binding = UploadListLayoutBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        swipeListRefreshLayout = binding.swipeContainingList;

        // this activity has no file really bound, it's for multiple accounts at the same time; should no inherit
        // from FileActivity; moreover, some behaviours inherited from FileActivity should be delegated to Fragments;
        // but that's other story
        setFile(null);

        // setup toolbar
        setupToolbar();

        // setup drawer
        setupDrawer(R.id.nav_uploads);

        setupContent();

        if (getSupportActionBar() != null) {
            ThemeUtils.setColoredTitle(getSupportActionBar(), R.string.uploads_view_title, this);
        }
    }

    private void setupContent() {
        binding.list.setEmptyView(binding.emptyList.getRoot());
        binding.emptyList.getRoot().setVisibility(View.GONE);
        binding.emptyList.emptyListProgress.setVisibility(View.GONE);
        binding.emptyList.emptyListIcon.setImageResource(R.drawable.uploads);
        binding.emptyList.emptyListIcon.getDrawable().mutate();
        binding.emptyList.emptyListIcon.setAlpha(0.5f);
        binding.emptyList.emptyListIcon.setVisibility(View.VISIBLE);
        binding.emptyList.emptyListViewHeadline.setText(getString(R.string.upload_list_empty_headline));
        binding.emptyList.emptyListViewText.setText(getString(R.string.upload_list_empty_text_auto_upload));
        binding.emptyList.emptyListViewText.setVisibility(View.VISIBLE);

        uploadListAdapter = new UploadListAdapter(this,
                                                  uploadsStorageManager,
                                                  getStorageManager(),
                                                  userAccountManager,
                                                  connectivityService,
                                                  powerManagementService,
                                                  clock);

        final GridLayoutManager lm = new GridLayoutManager(this, 1);
        uploadListAdapter.setLayoutManager(lm);

        int spacing = getResources().getDimensionPixelSize(R.dimen.media_grid_spacing);
        binding.list.addItemDecoration(new MediaGridItemDecoration(spacing));
        binding.list.setLayoutManager(lm);
        binding.list.setAdapter(uploadListAdapter);

        swipeListRefreshLayout.setOnRefreshListener(this::refresh);

        loadItems();
    }

    private void loadItems() {
        uploadListAdapter.loadUploadItemsFromDb();

        if (uploadListAdapter.getItemCount() > 0) {
            return;
        }

        swipeListRefreshLayout.setVisibility(View.VISIBLE);
        swipeListRefreshLayout.setRefreshing(false);
    }

    private void refresh() {
        // scan for missing auto uploads files
        Set<Job> jobs = JobManager.instance().getAllJobsForTag(FilesSyncJob.TAG);

        if (jobs.isEmpty()) {
            PersistableBundleCompat persistableBundleCompat = new PersistableBundleCompat();
            persistableBundleCompat.putBoolean(FilesSyncJob.OVERRIDE_POWER_SAVING, true);
            new JobRequest.Builder(FilesSyncJob.TAG)
                .setExact(1_000L)
                .setUpdateCurrent(false)
                .setExtras(persistableBundleCompat)
                .build()
                .schedule();
        }

        // retry failed uploads
        new Thread(() -> FileUploader.retryFailedUploads(
            this,
            null,
            uploadsStorageManager,
            connectivityService,
            userAccountManager,
            powerManagementService,
            null
        )).start();

        // update UI
        uploadListAdapter.loadUploadItemsFromDb();
        swipeListRefreshLayout.setRefreshing(false);
    }

    @Override
    protected void onStart() {
        super.onStart();
        ThemeUtils.setColoredTitle(getSupportActionBar(), R.string.uploads_view_title, this);
        final Optional<User> optionalUser = getUser();
        if (optionalUser.isPresent()) {
            setAccountInDrawer(optionalUser.get());
        }
    }

    @Override
    protected void onResume() {
        Log_OC.v(TAG, "onResume() start");
        super.onResume();

        setDrawerMenuItemChecked(R.id.nav_uploads);

        // Listen for upload messages
        uploadMessagesReceiver = new UploadMessagesReceiver();
        IntentFilter uploadIntentFilter = new IntentFilter();
        uploadIntentFilter.addAction(FileUploader.getUploadsAddedMessage());
        uploadIntentFilter.addAction(FileUploader.getUploadStartMessage());
        uploadIntentFilter.addAction(FileUploader.getUploadFinishMessage());
        registerReceiver(uploadMessagesReceiver, uploadIntentFilter);

        Log_OC.v(TAG, "onResume() end");

    }

    @Override
    protected void onPause() {
        Log_OC.v(TAG, "onPause() start");
        if (uploadMessagesReceiver != null) {
            unregisterReceiver(uploadMessagesReceiver);
            uploadMessagesReceiver = null;
        }
        super.onPause();
        Log_OC.v(TAG, "onPause() end");
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.activity_upload_list, menu);

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        boolean retval = true;
        switch (item.getItemId()) {
            case android.R.id.home:
                if (isDrawerOpen()) {
                    closeDrawer();
                } else {
                    openDrawer();
                }
                break;
            case R.id.action_clear_failed_uploads:
                uploadsStorageManager.clearFailedButNotDelayedUploads();
                uploadListAdapter.loadUploadItemsFromDb();
                break;

            default:
                retval = super.onOptionsItemSelected(item);
        }

        return retval;
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == FileActivity.REQUEST_CODE__UPDATE_CREDENTIALS && resultCode == RESULT_OK) {
            FilesSyncHelper.restartJobsIfNeeded(uploadsStorageManager,
                                                userAccountManager,
                                                connectivityService,
                                                powerManagementService);
        }
    }

    /**
     * @param operation Operation performed.
     * @param result    Result of the removal.
     */
    @Override
    public void onRemoteOperationFinish(RemoteOperation operation, RemoteOperationResult result) {
        if (operation instanceof CheckCurrentCredentialsOperation) {
            // Do not call super in this case; more refactoring needed around onRemoteOperationFinish :'(
            getFileOperationsHelper().setOpIdWaitingFor(Long.MAX_VALUE);
            dismissLoadingDialog();
            Account account = (Account) result.getData().get(0);
            if (!result.isSuccess()) {
                requestCredentialsUpdate(this, account);

            } else {
                // already updated -> just retry!
                FilesSyncHelper.restartJobsIfNeeded(uploadsStorageManager,
                                                    userAccountManager,
                                                    connectivityService,
                                                    powerManagementService);
            }

        } else {
            super.onRemoteOperationFinish(operation, result);
        }
    }


    @Override
    protected ServiceConnection newTransferenceServiceConnection() {
        return new UploadListServiceConnection();
    }

    /**
     * Defines callbacks for service binding, passed to bindService()
     */
    private class UploadListServiceConnection implements ServiceConnection {

        @Override
        public void onServiceConnected(ComponentName component, IBinder service) {
            if (service instanceof FileUploaderBinder) {
                if (mUploaderBinder == null) {
                    mUploaderBinder = (FileUploaderBinder) service;
                    Log_OC.d(TAG, "UploadListActivity connected to Upload service. component: " +
                            component + " service: " + service);
                } else {
                    Log_OC.d(TAG, "mUploaderBinder already set. mUploaderBinder: " +
                            mUploaderBinder + " service:" + service);
                }
            } else {
                Log_OC.d(TAG, "UploadListActivity not connected to Upload service. component: " +
                        component + " service: " + service);
            }
        }

        @Override
        public void onServiceDisconnected(ComponentName component) {
            if (component.equals(new ComponentName(UploadListActivity.this, FileUploader.class))) {
                Log_OC.d(TAG, "UploadListActivity suddenly disconnected from Upload service");
                mUploaderBinder = null;
            }
        }
    }

    /**
     * Once the file upload has changed its status -> update uploads list view
     */
    private class UploadMessagesReceiver extends BroadcastReceiver {
        /**
         * {@link BroadcastReceiver} to enable syncing feedback in UI
         */
        @Override
        public void onReceive(Context context, Intent intent) {
            try {
                uploadListAdapter.loadUploadItemsFromDb();
            } finally {
                if (intent != null) {
                    removeStickyBroadcast(intent);
                }
            }

        }
    }
}

File: app/src/main/java/com/owncloud/android/ui/activity/UploadListActivity.kt
/*
 * Nextcloud - Android Client
 *
 * SPDX-FileCopyrightText: 2026 Alper Ozturk <alper.ozturk@nextcloud.com>
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */
package com.owncloud.android.ui.activity

import android.accounts.Account
import android.annotation.SuppressLint
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.lifecycle.lifecycleScope
import androidx.localbroadcastmanager.content.LocalBroadcastManager
import androidx.recyclerview.widget.GridLayoutManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.google.android.material.snackbar.Snackbar
import com.nextcloud.client.account.User
import com.nextcloud.client.core.Clock
import com.nextcloud.client.device.PowerManagementService
import com.nextcloud.client.jobs.upload.FileUploadEventBroadcaster
import com.nextcloud.client.jobs.upload.FileUploadHelper
import com.nextcloud.client.utils.Throttler
import com.owncloud.android.R
import com.owncloud.android.databinding.UploadListLayoutBinding
import com.owncloud.android.datamodel.OCFile
import com.owncloud.android.datamodel.SyncedFolderProvider
import com.owncloud.android.datamodel.UploadsStorageManager
import com.owncloud.android.db.OCUpload
import com.owncloud.android.lib.common.operations.RemoteOperation
import com.owncloud.android.lib.common.operations.RemoteOperationResult
import com.owncloud.android.lib.common.utils.Log_OC
import com.owncloud.android.operations.CheckCurrentCredentialsOperation
import com.owncloud.android.ui.adapter.uploadList.UploadListAdapter
import com.owncloud.android.ui.adapter.uploadList.helper.ConflictHandlingResult
import com.owncloud.android.ui.adapter.uploadList.helper.UploadListAdapterAction
import com.owncloud.android.ui.adapter.uploadList.helper.UploadListAdapterActionHandler
import com.owncloud.android.ui.adapter.uploadList.helper.UploadListAdapterHelper
import com.owncloud.android.ui.adapter.uploadList.helper.UploadListItemOnClick
import com.owncloud.android.ui.decoration.MediaGridItemDecoration
import com.owncloud.android.utils.DisplayUtils
import com.owncloud.android.utils.FilesSyncHelper
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import javax.inject.Inject

@Suppress("MagicNumber")
class UploadListActivity :
    FileActivity(),
    UploadListItemOnClick {
    @Inject lateinit var uploadsStorageManager: UploadsStorageManager

    @Inject lateinit var powerManagementService: PowerManagementService

    @Inject lateinit var clock: Clock

    @Inject lateinit var syncedFolderProvider: SyncedFolderProvider

    @Inject lateinit var localBroadcastManager: LocalBroadcastManager

    @Inject lateinit var throttler: Throttler

    private var swipeListRefreshLayout: SwipeRefreshLayout? = null
    private var binding: UploadListLayoutBinding? = null

    private var uploadFinishReceiver: UploadFinishReceiver? = null
    private lateinit var uploadListAdapter: UploadListAdapter
    private lateinit var adapterActionHandler: UploadListAdapterAction
    private lateinit var adapterHelper: UploadListAdapterHelper

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        throttler.intervalMillis = 1000
        binding = UploadListLayoutBinding.inflate(layoutInflater)
        val binding = binding!!
        setContentView(binding.getRoot())
        swipeListRefreshLayout = binding.swipeContainingList

        // this activity has no file really bound, it's for multiple accounts at the same time; should no inherit
        // from FileActivity; moreover, some behaviours inherited from FileActivity should be delegated to Fragments;
        // but that's other story
        file = null

        setupToolbar()
        updateActionBarTitleAndHomeButtonByString(getString(R.string.uploads_view_title))
        setupDrawer(menuItemId)
        setupContent()
    }

    override fun getMenuItemId() = R.id.nav_uploads

    private fun setupContent() {
        setupEmptyList()
        adapterActionHandler = UploadListAdapterActionHandler()
        adapterHelper = UploadListAdapterHelper(this)
        uploadListAdapter = UploadListAdapter(
            this,
            uploadsStorageManager,
            userAccountManager,
            connectivityService,
            powerManagementService,
            viewThemeUtils,
            this,
            adapterHelper
        )

        val lm = GridLayoutManager(this, 1)
        uploadListAdapter.setLayoutManager(lm)

        val spacing = getResources().getDimensionPixelSize(R.dimen.media_grid_spacing)
        binding?.list?.run {
            addItemDecoration(MediaGridItemDecoration(spacing))
            setLayoutManager(lm)
            setAdapter(uploadListAdapter)
        }

        swipeListRefreshLayout?.let { viewThemeUtils.androidx.themeSwipeRefreshLayout(it) }
        swipeListRefreshLayout?.setOnRefreshListener { this.refresh() }
        loadItems()
    }

    private fun setupEmptyList() {
        binding?.run {
            list.setEmptyView(emptyList.getRoot())
            emptyList.run {
                root.visibility = View.GONE

                emptyListIcon.run {
                    setImageResource(R.drawable.uploads)
                    getDrawable().mutate()
                    setAlpha(0.5f)
                    setVisibility(View.VISIBLE)
                }

                emptyListViewHeadline.text = getString(R.string.upload_list_empty_headline)

                emptyListViewText.run {
                    text = getString(R.string.upload_list_empty_text_auto_upload)
                    visibility = View.VISIBLE
                }
            }
        }
    }

    private fun loadItems() {
        swipeListRefreshLayout?.isRefreshing = true
        uploadListAdapter.loadUploadItemsFromDb { swipeListRefreshLayout?.isRefreshing = false }
    }

    private fun refresh() {
        val isUploadStarted = FileUploadHelper.instance().retryFailedUploads(
            uploadsStorageManager,
            connectivityService,
            accountManager,
            powerManagementService
        )

        if (!isUploadStarted) {
            uploadListAdapter.loadUploadItemsFromDb { swipeListRefreshLayout?.isRefreshing = false }
        }
    }

    override fun onStart() {
        Log_OC.v(TAG, "onStart() start")
        super.onStart()

        highlightNavigationViewItem(menuItemId)

        uploadFinishReceiver = UploadFinishReceiver()
        val intentFilter = IntentFilter().apply {
            addAction(FileUploadEventBroadcaster.ACTION_UPLOAD_ENQUEUED)
            addAction(FileUploadEventBroadcaster.ACTION_UPLOAD_STARTED)
            addAction(FileUploadEventBroadcaster.ACTION_UPLOAD_COMPLETED)
        }
        uploadFinishReceiver?.let { localBroadcastManager.registerReceiver(it, intentFilter) }

        Log_OC.v(TAG, "onStart() end")
    }

    override fun onStop() {
        Log_OC.v(TAG, "onStop() start")
        if (uploadFinishReceiver != null) {
            uploadFinishReceiver?.let { localBroadcastManager.unregisterReceiver(it) }
            uploadFinishReceiver = null
        }
        super.onStop()
        Log_OC.v(TAG, "onStop() end")
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.activity_upload_list, menu)
        menu.findItem(R.id.action_toggle_global_pause)?.let { updateGlobalPauseIcon(it) }
        return true
    }

    private fun updateGlobalPauseIcon(item: MenuItem) {
        val paused = preferences.isGlobalUploadPaused()
        item.setIcon(if (paused) R.drawable.ic_global_resume else R.drawable.ic_global_pause)
        item.title = getString(
            if (paused) {
                R.string.upload_action_global_upload_resume
            } else {
                R.string.upload_action_global_upload_pause
            }
        )
    }

    @SuppressLint("NotifyDataSetChanged")
    private fun toggleGlobalPause(item: MenuItem) {
        preferences.setGlobalUploadPaused(!preferences.isGlobalUploadPaused())
        updateGlobalPauseIcon(item)
        val uploadHelper = FileUploadHelper.instance()
        accountManager.getAllUsers().filterNotNull().forEach { user ->
            val ids = uploadsStorageManager.getCurrentUploadIds(user.accountName)
            uploadHelper.cancelAndRestartUploadJob(user, ids)
        }
        uploadListAdapter.notifyDataSetChanged()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean = when (item.itemId) {
        android.R.id.home -> {
            if (isDrawerOpen) closeDrawer() else openDrawer()
            true
        }

        R.id.action_toggle_global_pause -> {
            toggleGlobalPause(item)
            true
        }

        else -> super.onOptionsItemSelected(item)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == REQUEST_CODE__UPDATE_CREDENTIALS && resultCode == RESULT_OK) {
            FilesSyncHelper.restartUploadsIfNeeded(
                uploadsStorageManager,
                userAccountManager,
                connectivityService,
                powerManagementService
            )
        }
    }

    override fun onRemoteOperationFinish(operation: RemoteOperation<*>?, result: RemoteOperationResult<*>) {
        if (operation !is CheckCurrentCredentialsOperation) {
            super.onRemoteOperationFinish(operation, result)
            return
        }

        fileOperationsHelper.opIdWaitingFor = Long.MAX_VALUE
        dismissLoadingDialog()
        val account = result.data[0] as? Account
        if (!result.isSuccess) {
            requestCredentialsUpdate(account)
        } else {
            FilesSyncHelper.restartUploadsIfNeeded(
                uploadsStorageManager,
                userAccountManager,
                connectivityService,
                powerManagementService
            )
        }
    }

    override fun onLastUploadResultConflictClick(upload: OCUpload) {
        DisplayUtils.showSnackMessage(this, R.string.upload_sync_conflict_checking)

        lifecycleScope.launch {
            val client = clientRepository.getOwncloudClient() ?: return@launch
            val result = adapterActionHandler.handleConflict(upload, client, uploadsStorageManager)

            withContext(Dispatchers.Main) {
                when (result) {
                    is ConflictHandlingResult.ConflictNotExists -> {
                        uploadListAdapter.notifyUploadChanged(upload)
                        onConflictNotExists(upload)
                    }

                    is ConflictHandlingResult.CannotCheckConflict -> {
                        DisplayUtils.showSnackMessage(
                            this@UploadListActivity,
                            R.string.upload_sync_conflict_check_error
                        )
                    }

                    is ConflictHandlingResult.ShowConflictResolveDialog -> {
                        adapterHelper.openConflictActivity(result.file, result.upload)
                    }
                }
            }
        }
    }

    private fun onConflictNotExists(upload: OCUpload) {
        val rootView = binding?.root ?: return
        val snackbar = Snackbar.make(
            rootView,
            R.string.upload_sync_conflict_not_exists,
            Snackbar.LENGTH_LONG
        )

        snackbar.setAction(R.string.retry) {
            val optionalUser = userAccountManager.getUser(upload.accountName)
            if (optionalUser.isPresent) {
                FileUploadHelper.instance().retryUpload(upload, optionalUser.get())
            }
        }

        snackbar.show()
    }

    private inner class UploadFinishReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            throttler.run("update_upload_list") { uploadListAdapter.loadUploadItemsFromDb() }
        }
    }

    companion object {
        private val TAG: String = UploadListActivity::class.java.getSimpleName()

        fun createIntent(file: OCFile?, user: User?, flag: Int?, context: Context?): Intent =
            Intent(context, UploadListActivity::class.java).apply {
                if (flag != null) {
                    setFlags(flags or flag)
                }
                putExtra(EXTRA_FILE, file)
                putExtra(EXTRA_USER, user)
            }
    }
}
